cas-server-support-jpa-ticket-registry/src/main/java/org/jasig/cas/ticket/registry/JpaTicketRegistry.java

```diff
@@ -211,6 +165,61 @@ public int serviceTicketCount() {
         return countToInt(entityManager.createQuery("select count(t) from ServiceTicketImpl t").getSingleResult());
     }
 
+    @Override
+    public boolean deleteSingleTicket(final String ticketId) {
+        final Ticket ticket = getTicket(ticketId);
+        int failureCount = 0;
+
+        if (ticket instanceof ServiceTicket) {
+            failureCount = deleteServiceTickets(ticketId);
+        } else if (ticket instanceof TicketGrantingTicket) {
+            failureCount = deleteTicketGrantingTickets(ticketId);
+        } else {
+            throw new IllegalArgumentException("Invalid ticket type with id " + ticketId);
+        }
+        return failureCount == 0;
+    }
+
+    <T extends Ticket> List<T> getTicketQueryResultList(final String ticketId, final String query, final Class<? extends Ticket> clazz) {
+        return (List) entityManager.createQuery(query, clazz)
+                .setParameter("id", ticketId)
+                .getResultList();
+    }
+
+    private int deleteServiceTickets(final String ticketId) {
+        final List<ServiceTicketImpl> serviceTicketImpls = getTicketQueryResultList(ticketId,
+                "select s from ServiceTicketImpl s where s.id = :id", ServiceTicketImpl.class);
+        return deleteTicketsFromResultList(serviceTicketImpls);
+    }
+
+    private int deleteTicketsFromResultList(final List<? extends Ticket> serviceTicketImpls) {
+        int failureCount = 0;
+        for (final Ticket serviceTicketImpl : serviceTicketImpls) {
+            if (!removeTicket(serviceTicketImpl)) {
+                failureCount++;
+            }
+        }
+        return failureCount;
+    }
+
+    private int deleteTicketGrantingTickets(final String ticketId) {
+        int failureCount = 0;
+
+        final List<ServiceTicketImpl> serviceTicketImpls = getTicketQueryResultList(ticketId,
+                "select s from ServiceTicketImpl s where s.ticketGrantingTicket.id = :id", ServiceTicketImpl.class);
+        failureCount += deleteTicketsFromResultList(serviceTicketImpls);
+
+        List<TicketGrantingTicketImpl> ticketGrantingTicketImpls = getTicketQueryResultList(ticketId,
+                "select t from TicketGrantingTicketImpl t where t.ticketGrantingTicket.id = :id", TicketGrantingTicketImpl.class);
+        failureCount += deleteTicketsFromResultList(ticketGrantingTicketImpls);
+
+        ticketGrantingTicketImpls = getTicketQueryResultList(ticketId,
+                "select t from TicketGrantingTicketImpl t where t.id = :id", TicketGrantingTicketImpl.class);
+        failureCount += deleteTicketsFromResultList(ticketGrantingTicketImpls);
+
+        return failureCount;
```

Comment:
JPA is tricky because we need preserve transaction integrity. So we clean up everything in one pass, allowing further ops to become nops.